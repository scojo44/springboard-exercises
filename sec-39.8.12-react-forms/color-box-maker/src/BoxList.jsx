import React, {useState} from "react";
import {v4 as uuid} from 'uuid';
import NewBoxForm from "./NewBoxForm";
import Box from './Box';

const BoxList = () => {
  function addBox({color, width, height}) {
    setBoxes(() => [...boxes, {id: uuid(), color, width, height}]);
  }

  function removeBox(id) {
    setBoxes(() => boxes.filter(box => box.id !== id));
  }

  const [boxes, setBoxes] = useState([]);
  const boxLIs = boxes.map(({id, width, height, color}) => (
    <Box id={id} width={width} height={height} color={color} key={id} removeMe={removeBox} />
  ));

  return (
    <div className="BoxList">
      <h1>Color Box Maker</h1>
      <NewBoxForm addBox={addBox} />
      <ul>{boxLIs}</ul>
    </div>
  )
};

export default BoxList;
