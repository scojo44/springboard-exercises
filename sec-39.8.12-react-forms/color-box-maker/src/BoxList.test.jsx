import { render, fireEvent } from '@testing-library/react';
import BoxList from './App';

describe('BoxList Tests', () => {
  it('renders without crashing', () => {
    render(<BoxList />);
  });

  it('matches snapshot', () => {
    const {asFragment} = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('BoxList - Add/Remove Box Tests', () => {
  function createBox() {
    const {container, getByText, getByLabelText} = render(<BoxList />);
    const color = getByLabelText('Color:');
    const width = getByLabelText('Width:');
    const height = getByLabelText('Height:');
    const button = getByText('Add Box');

    // Confirm no boxes displayed
    expect(container.querySelectorAll('.Box').length).toBe(0);

    // Create a box
    fireEvent.change(color, {target: {value: 'thistle'}});
    fireEvent.change(width, {target: {value: '100px'}});
    fireEvent.change(height, {target: {value: '100px'}});
    fireEvent.click(button);

    return {container, getByText, getByLabelText};
  }

  it('creates a new box and adds it to the list', () => {
    const {container} = createBox();
    const box = container.querySelector('.Box div');
    expect(container.querySelectorAll('.Box').length).toBe(1);
    expect(container.querySelector('.BoxList ul')).toContainElement(box);
    expect(box).toHaveStyle({backgroundColor: '#D8BFD8'}); // thistle
    expect(box).toHaveStyle({width: "100px"});
    expect(box).toHaveStyle({height: "100px"});
  })

  it('removes the new box when the X button is clicked', () => {
    const {container, getByText} = createBox();
    const box = container.querySelector('.Box div');
    const removeButton = getByText('X');
    fireEvent.click(removeButton);
    expect(container.querySelectorAll('.Box').length).toBe(0);
  });
});
