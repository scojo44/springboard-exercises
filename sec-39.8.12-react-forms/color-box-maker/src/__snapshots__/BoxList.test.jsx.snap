// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BoxList Tests > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <h1>
      Color Box Maker
    </h1>
    <form
      class="NewBoxForm"
    >
      <p>
        <label
          for="color"
        >
          Color: 
        </label>
        <input
          id="color"
          name="color"
          type="text"
          value=""
        />
      </p>
      <p>
        <label
          for="width"
        >
          Width: 
        </label>
        <input
          id="width"
          name="width"
          type="text"
          value=""
        />
      </p>
      <p>
        <label
          for="height"
        >
          Height: 
        </label>
        <input
          id="height"
          name="height"
          type="text"
          value=""
        />
      </p>
      <button
        type="submit"
      >
        Add Box
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
